File: src/bin/agent-cli.rs
use std::{
    io::{self, Write},
    time::Duration,
};

use anyhow::{Context, Result};
use clap::{Parser, Subcommand};
use reqwest::Client;
use rust_ai_agent::{
    cliproxy::{CliProxyLoginStart, CliProxyLoginStatus, CliProxyModel},
    model::{AgentRequest, AgentResponse},
};
use serde::{Deserialize, Serialize};
use tokio::time::sleep;
use uuid::Uuid;

#[derive(Debug, Parser)]
#[command(name = "agent-cli", version, about = "Rust AI Agent 终端客户端")]
struct Cli {
    /// Rust Agent HTTP 服务地址
    #[arg(
        long,
        env = "AGENT_SERVER_URL",
        default_value = "http://127.0.0.1:8080"
    )]
    server: String,

    /// 当前用户，用于长期记忆隔离
    #[arg(long, env = "AGENT_USER_ID", default_value = "local-user")]
    user_id: String,

    /// 当前会话 ID；不指定时自动生成
    #[arg(long, env = "AGENT_SESSION_ID")]
    session_id: Option<String>,

    #[command(subcommand)]
    command: Option<Command>,
}

#[derive(Debug, Subcommand)]
enum Command {
    /// 进入连续聊天模式；传入 prompt 时执行一次后退出
    Chat { prompt: Option<String> },
    /// 执行一次 Agent 请求
    Run {
        input: String,
        #[arg(long)]
        session_id: Option<String>,
    },
    /// 检查 Rust Agent 服务状态
    Health,
    /// 查看 CLIProxyAPI 可用模型
    Models,
    /// 验证 CLIProxyAPI 和指定模型
    Verify {
        #[arg(long)]
        model: Option<String>,
    },
    /// 发起 CLIProxyAPI OAuth 登录
    Login {
        provider: String,
        #[arg(long, default_value_t = false)]
        wait: bool,
    },
}

#[derive(Clone)]
struct AgentApi {
    client: Client,
    base_url: String,
}

#[derive(Debug, Deserialize, Serialize)]
struct HealthResponse {
    status: String,
    provider: String,
    agents: Vec<String>,
    sandbox_enabled: bool,
    cliproxyapi_configured: bool,
}

#[derive(Debug, Serialize)]
struct VerifyRequest {
    model: Option<String>,
}

#[derive(Debug, Deserialize, Serialize)]
struct VerifyResponse {
    reachable: bool,
    model_count: usize,
    models: Vec<CliProxyModel>,
    selected_model: Option<String>,
    selected_model_available: Option<bool>,
}

impl AgentApi {
    fn new(server: String) -> Self {
        Self {
            client: Client::new(),
            base_url: server.trim_end_matches('/').to_string(),
        }
    }

    async fn run(&self, request: AgentRequest) -> Result<AgentResponse> {
        self.client
            .post(self.url("/v1/agent/run"))
            .json(&request)
            .send()
            .await
            .context("无法连接 Rust Agent 服务")?
            .error_for_status()
            .context("Agent 请求失败")?
            .json()
            .await
            .context("无法解析 Agent 响应")
    }

    async fn health(&self) -> Result<HealthResponse> {
        self.client
            .get(self.url("/health"))
            .send()
            .await
            .context("无法连接 Rust Agent 服务")?
            .error_for_status()?
            .json()
            .await
            .context("无法解析健康检查响应")
    }

    async fn models(&self) -> Result<Vec<CliProxyModel>> {
        self.client
            .get(self.url("/v1/providers/cliproxyapi/models"))
            .send()
            .await
            .context("无法连接 CLIProxyAPI 模型接口")?
            .error_for_status()
            .context("查询 CLIProxyAPI 模型失败")?
            .json()
            .await
            .context("无法解析模型列表")
    }

    async fn verify(&self, model: Option<String>) -> Result<VerifyResponse> {
        self.client
            .post(self.url("/v1/providers/cliproxyapi/verify"))
            .json(&VerifyRequest { model })
            .send()
            .await
            .context("无法连接 CLIProxyAPI 验证接口")?
            .error_for_status()
            .context("CLIProxyAPI 验证失败")?
            .json()
            .await
            .context("无法解析验证响应")
    }

    async fn login(&self, provider: &str) -> Result<CliProxyLoginStart> {
        self.client
            .post(self.url("/v1/providers/cliproxyapi/login"))
            .json(&serde_json::json!({ "provider": provider }))
            .send()
            .await
            .context("无法连接 CLIProxyAPI 登录接口")?
            .error_for_status()
            .context("CLIProxyAPI 登录启动失败")?
            .json()
            .await
            .context("无法解析登录响应")
    }

    async fn login_status(&self, state: &str) -> Result<CliProxyLoginStatus> {
        self.client
            .get(self.url("/v1/providers/cliproxyapi/login/status"))
            .query(&[("state", state)])
            .send()
            .await
            .context("无法连接 CLIProxyAPI 登录状态接口")?
            .error_for_status()
            .context("查询 CLIProxyAPI 登录状态失败")?
            .json()
            .await
            .context("无法解析登录状态响应")
    }

    fn url(&self, path: &str) -> String {
        format!("{}{}", self.base_url, path)
    }
}

#[tokio::main]
async fn main() -> Result<()> {
    let cli = Cli::parse();
    let api = AgentApi::new(cli.server);
    let default_session = cli.session_id.unwrap_or_else(|| Uuid::new_v4().to_string());

    match cli.command.unwrap_or(Command::Chat { prompt: None }) {
        Command::Chat {
            prompt: Some(prompt),
        } => {
            print_response(
                &api.run(request(&default_session, &cli.user_id, prompt))
                    .await?,
            );
        }
        Command::Chat { prompt: None } => {
            interactive_chat(&api, &cli.user_id, &default_session).await?
        }
        Command::Run { input, session_id } => {
            let session_id = session_id.as_deref().unwrap_or(&default_session);
            print_response(&api.run(request(session_id, &cli.user_id, input)).await?);
        }
        Command::Health => {
            let health = api.health().await?;
            println!("{}", serde_json::to_string_pretty(&health)?);
        }
        Command::Models => {
            let models = api.models().await?;
            for model in models {
                println!(
                    "{}{}",
                    model.id,
                    model
                        .owned_by
                        .map(|v| format!(" ({v})"))
                        .unwrap_or_default()
                );
            }
        }
        Command::Verify { model } => {
            println!(
                "{}",
                serde_json::to_string_pretty(&api.verify(model).await?)?
            );
        }
        Command::Login { provider, wait } => login_flow(&api, &provider, wait).await?,
    }
    Ok(())
}

async fn interactive_chat(api: &AgentApi, user_id: &str, session_id: &str) -> Result<()> {
    println!("Rust AI Agent CLI | session={session_id}");
    println!("输入消息开始对话，输入 /help 查看命令，输入 /exit 退出。\n");
    let stdin = io::stdin();
    loop {
        print!("you> ");
        io::stdout().flush()?;
        let mut line = String::new();
        stdin.read_line(&mut line)?;
        let input = line.trim();
        match input {
            "" => continue,
            "/exit" | "/quit" => break,
            "/help" => {
                println!(
                    "/exit 退出；/health 检查服务；/models 查看模型；其他文本发送给 Agent。\n"
                );
            }
            "/health" => println!("{}", serde_json::to_string_pretty(&api.health().await?)?),
            "/models" => {
                for model in api.models().await? {
                    println!("- {}", model.id);
                }
            }
            message => {
                print!("agent> ");
                io::stdout().flush()?;
                let response = api
                    .run(request(session_id, user_id, message.to_string()))
                    .await?;
                println!("{}\n", response.output);
                println!(
                    "plan: {} steps | reflection: {}\n",
                    response.plan.steps.len(),
                    response.reflection.passed
                );
            }
        }
    }
    Ok(())
}

async fn login_flow(api: &AgentApi, provider: &str, wait: bool) -> Result<()> {
    let login = api.login(provider).await?;
    println!("provider: {provider}");
    println!("state: {}", login.state.as_deref().unwrap_or("<missing>"));
    println!(
        "请在浏览器打开 OAuth URL：\n{}",
        login.url.as_deref().unwrap_or("<missing>")
    );
    let Some(state) = login.state else {
        return Ok(());
    };
    if !wait {
        println!("登录后执行：agent-cli --server ... login {provider} --wait 不会复用本次 state，请使用 HTTP status 接口或桌面版轮询。 ");
        return Ok(());
    }

    loop {
        sleep(Duration::from_secs(2)).await;
        let status = api.login_status(&state).await?;
        println!(
            "status={} authenticated={}",
            status.status, status.authenticated
        );
        if status.status != "wait" {
            if let Some(error) = status.error {
                println!("error: {error}");
            }
            break;
        }
    }
    Ok(())
}

fn request(session_id: &str, user_id: &str, input: String) -> AgentRequest {
    AgentRequest {
        session_id: session_id.to_string(),
        user_id: Some(user_id.to_string()),
        model: None,
        input,
    }
}

fn print_response(response: &AgentResponse) {
    println!("{}", response.output);
    println!(
        "\n[session={} execution={} score={:.1}]",
        response.session_id, response.execution_id, response.evaluation.total_score
    );
}

File: packaging/npm/agent-cli/bin/agent-cli.js
#!/usr/bin/env node

const readline = require("node:readline/promises");
const { stdin, stdout } = require("node:process");

const defaultServer = process.env.AGENT_SERVER_URL || "http://127.0.0.1:8080";
const defaultUser = process.env.AGENT_USER_ID || "local-user";

function usage() {
  console.log(`Rust AI Agent npm CLI

Usage:
  agent-cli [options] chat [prompt]
  agent-cli [options] run <input>
  agent-cli [options] health
  agent-cli [options] models
  agent-cli [options] verify [--model <model>]
  agent-cli [options] login <provider> [--wait]
  agent-cli [options] sessions
  agent-cli [options] session <id>

Options:
  --server <url>       Rust Agent URL (default: ${defaultServer})
  --user-id <id>       User ID (default: ${defaultUser})
  --session-id <id>    Session ID
  --mode <mode>        Permission mode: default / plan / acceptEdits / bypassPermissions / dontAsk
  -h, --help           Show that help
`);
}

function parseArgs(argv) {
  const options = {
    server: defaultServer,
    userId: defaultUser,
    sessionId: process.env.AGENT_SESSION_ID || cryptoRandomId(),
    command: null,
    args: [],
    wait: false,
    model: null,
    mode: process.env.AGENT_PERMISSION_MODE || null,
  };
  let index = 0;
  while (index < argv.length) {
    const value = argv[index];
    if (value === "--server") options.server = argv[++index];
    else if (value === "--user-id") options.userId = argv[++index];
    else if (value === "--session-id") options.sessionId = argv[++index];
    else if (value === "--model") options.model = argv[++index];
    else if (value === "--wait") options.wait = true;
    else if (value === "--mode") options.mode = argv[++index];
    else if (value === "-h" || value === "--help") options.help = true;
    else if (!options.command) options.command = value;
    else options.args.push(value);
    index += 1;
  }
  return options;
}

function cryptoRandomId() {
  return `npm-${Date.now()}-${Math.random().toString(16).slice(2)}`;
}

function endpoint(server, path) {
  return `${server.replace(/\/$/, "")}${path}`;
}

async function request(options, path, init = {}) {
  let response;
  try {
    response = await fetch(endpoint(options.server, path), {
      ...init,
      headers: { "content-type": "application/json", ...(init.headers || {}) },
    });
  } catch (error) {
    throw new Error(`无法连接 Rust Agent 服务：${error.message}`);
  }
  const text = await response.text();
  let body;
  try {
    body = text ? JSON.parse(text) : {};
  } catch {
    body = { raw: text };
  }
  if (!response.ok) {
    throw new Error(body.error || `HTTP ${response.status}`);
  }
  return body;
}

async function runAgent(options, input) {
  const body = {
    session_id: options.sessionId,
    user_id: options.userId,
    input,
  };
  if (options.mode) body.mode = options.mode;
  return request(options, "/v1/agent/run", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

function renderTodos(todos) {
  if (!Array.isArray(todos) || !todos.length) return "";
  const lines = todos.map((todo, index) => {
    const active = todo.active_form || todo.content;
    return `${index + 1}. [${todo.status}] ${active}`;
  });
  return `\ntodos:\n${lines.map((line) => `  ${line}`).join("\n")}\n`;
}

function printResponse(response) {
  console.log(response.output);
  console.log(renderTodos(response.todos));
  console.log(`[session=${response.session_id} execution=${response.execution_id} score=${Number(response.evaluation?.total_score || 0).toFixed(1)} turns=${response.turns ?? 0} tool_calls=${response.tool_calls ?? 0} tokens=${response.usage?.input_tokens ?? 0}in/${response.usage?.output_tokens ?? 0}out]`);
}

async function chat(options, prompt) {
  if (prompt) {
    printResponse(await runAgent(options, prompt));
    return;
  }
  const rl = readline.createInterface({ input: stdin, output: stdout });
  console.log(`Rust AI Agent npm CLI | session=${options.sessionId}`);
  console.log("输入消息开始对话，输入 /help 查看命令，输入 /exit 退出。\n");
  try {
    while (true) {
      const input = (await rl.question("you> ")).trim();
      if (!input) continue;
      if (input === "/exit" || input === "/quit") break;
      if (input === "/help") {
        console.log("/exit 退出；/health 检查服务；/models 查看模型；/sessions 列出会话；其他文本发送给 Agent。\n");
        continue;
      }
      if (input === "/health") {
        console.log(JSON.stringify(await request(options, "/health"), null, 2));
        continue;
      }
      if (input === "/models") {
        const models = await request(options, "/v1/providers/cliproxyapi/models");
        console.log(models.map((model) => `- ${model.id}`).join("\n"));
        continue;
      }
      if (input === "/sessions") {
        const sessions = await request(options, "/v1/sessions");
        for (const session of sessions) {
          console.log(`${session.id}  messages=${session.message_count}  updated=${session.updated_at}`);
        }
        continue;
      }
      printResponse(await runAgent(options, input));
    }
  } finally {
    rl.close();
  }
}

async function login(options, provider) {
  const started = await request(options, "/v1/providers/cliproxyapi/login", {
    method: "POST",
    body: JSON.stringify({ provider }),
  });
  console.log(`provider: ${provider}`);
  console.log(`state: ${started.state || "<missing>"}`);
  console.log(`OAuth URL:\n${started.url || "<missing>"}`);
  if (!options.wait || !started.state) return;
  while (true) {
    await new Promise((resolve) => setTimeout(resolve, 2000));
    const status = await request(options, `/v1/providers/cliproxyapi/login/status?state=${encodeURIComponent(started.state)}`);
    console.log(`status=${status.status} authenticated=${status.authenticated}`);
    if (status.status !== "wait") break;
  }
}

async function main() {
  const options = parseArgs(process.argv.slice(2));
  if (options.help) {
    usage();
    return;
  }
  if (!options.command) {
    await chat(options);
    return;
  }
  switch (options.command) {
    case "chat":
      await chat(options, options.args.join(" "));
      break;
    case "run":
      if (!options.args.length) throw new Error("run 需要输入任务文本");
      printResponse(await runAgent(options, options.args.join(" ")));
      break;
    case "health":
      console.log(JSON.stringify(await request(options, "/health"), null, 2));
      break;
    case "models":
      console.log(JSON.stringify(await request(options, "/v1/providers/cliproxyapi/models"), null, 2));
      break;
    case "verify":
      console.log(JSON.stringify(await request(options, "/v1/providers/cliproxyapi/verify", {
        method: "POST",
        body: JSON.stringify({ model: options.model }),
      }), null, 2));
      break;
    case "login":
      if (!options.args[0]) throw new Error("login 需要 provider，例如 codex");
      await login(options, options.args[0]);
      break;
    case "sessions": {
      const sessions = await request(options, "/v1/sessions");
      if (!sessions.length) console.log("（暂无会话）");
      for (const session of sessions) {
        console.log(`${session.id}  messages=${session.message_count}  updated=${session.updated_at}`);
      }
      break;
    }
    case "session": {
      if (!options.args[0]) throw new Error("session 需要会话 ID");
      const session = await request(options, `/v1/sessions/${encodeURIComponent(options.args[0])}`);
      console.log(`session: ${session.id}`);
      console.log(`tokens: ${session.usage?.input_tokens ?? 0} in / ${session.usage?.output_tokens ?? 0} out`);
      if (Array.isArray(session.todos) && session.todos.length) {
        console.log("todos:");
        session.todos.forEach((todo, index) => {
          console.log(`  ${index + 1}. [${todo.status}] ${todo.active_form || todo.content}`);
        });
      }
      break;
    }
    default:
      throw new Error(`未知命令：${options.command}`);
  }
}

main().catch((error) => {
  console.error(`错误：${error.message}`);
  process.exitCode = 1;
});
